packages/ws/src/strategies/sharding/WorkerShardingStrategy.ts

```diff
@@ -0,0 +1,212 @@
+import type { WebSocketShardDestroyOptions, WebSocketShardEvents } from '../../ws/WebSocketShard';
+import { FetchingStrategyOptions, managerToFetchingStrategyOptions } from '../context/IContextFetchingStrategy';
+
+export interface WorkerData extends FetchingStrategyOptions {
+	shardIds: number[];
+}
+
+export enum WorkerSendPayloadOp {
+	connect,
+	destroy,
+	send,
+	sessionInfoResponse,
+}
+
+export type WorkerSendPayload =
+	| { op: WorkerSendPayloadOp.connect; shardId: number }
+	| { op: WorkerSendPayloadOp.destroy; shardId: number; options?: WebSocketShardDestroyOptions }
+	| { op: WorkerSendPayloadOp.send; shardId: number; payload: GatewaySendPayload }
+	| { op: WorkerSendPayloadOp.sessionInfoResponse; nonce: number; session: SessionInfo | null };
+
+export enum WorkerRecievePayloadOp {
+	connected,
+	destroyed,
+	event,
+	retrieveSessionInfo,
+	updateSessionInfo,
+}
+
+export type WorkerRecievePayload =
+	| { op: WorkerRecievePayloadOp.connected; shardId: number }
+	| { op: WorkerRecievePayloadOp.destroyed; shardId: number }
+	// Can't seem to get a type-safe union based off of the event, so I'm sadly leaving data as any for now
+	| { op: WorkerRecievePayloadOp.event; shardId: number; event: WebSocketShardEvents; data: any }
+	| { op: WorkerRecievePayloadOp.retrieveSessionInfo; shardId: number; nonce: number }
+	| { op: WorkerRecievePayloadOp.updateSessionInfo; shardId: number; session: SessionInfo | null };
+
+/**
+ * Options for a {@link WorkerShardingStrategy}
+ */
+export interface WorkerShardingStrategyOptions {
+	/**
+	 * Dictates how many shards should be spawned per worker thread.
+	 */
+	shardsPerWorker: number | 'all';
+}
+
+/**
+ * Strategy used to spawn threads in worker_threads
+ */
+export class WorkerShardingStrategy implements IShardingStrategy {
+	private readonly manager: WebSocketManager;
+	private readonly options: WorkerShardingStrategyOptions;
+
+	private workers: Worker[] = [];
+	private readonly workerByShardId = new Collection<number, Worker>();
+
+	private readonly connectPromises = new Collection<number, () => void>();
+	private readonly destroyPromises = new Collection<number, () => void>();
+
+	private readonly throttler: IdentifyThrottler;
+
+	public constructor(manager: WebSocketManager, options: WorkerShardingStrategyOptions) {
+		this.manager = manager;
+		this.throttler = new IdentifyThrottler(manager);
+		this.options = options;
+	}
+
+	public async spawn(shardIds: number[]) {
+		await this.destroy({ reason: 'User is adjusting their shards' });
+		const shardsPerWorker = this.options.shardsPerWorker === 'all' ? shardIds.length : this.options.shardsPerWorker;
+
+		let shards = 0;
+		while (shards !== shardIds.length) {
+			const slice = shardIds.slice(shards, shardsPerWorker + shards);
+			const workerData: WorkerData = {
+				...(await managerToFetchingStrategyOptions(this.manager)),
+				shardIds: slice,
+			};
+
+			const worker = new Worker(join(__dirname, 'worker'), { workerData });
```

Comment:
No extension?

Reply:
Seems to work without, should I add it anyways?

Reply:
Yeah